Bug 2076619: Modified git import flow module to handle create button enable-disable issue

lokanandaprabhu opened the pull request:
**Fixes**: 
https://issues.redhat.com/browse/OCPBUGSM-43404

**Analysis / Root cause**: 
1. Build starategy was not update when builder image is selected by default, so the create button was not getting enabled.
2. No separate message when devfile is not detected for Git type - other
3. On change of import strategy error message was not cleared.

**Solution Description**: 
1. Updated Build starategy to 'Source'.
2. Disabled Devfile import strategy for Git type - other
3. On change of import strategy calling a method which will handle the error showing part.

**Screen shots / Gifs for design review**: 

https://user-images.githubusercontent.com/102503482/191943839-30293edb-8866-4867-80ae-3e1b358689fa.mov


**Unit test coverage report**: 
NA

**Test setup:**
CASE A
1. Switch to developer perspective, add page > import from Git
2. Enter URL https://mysupersecretgit.example.com/org/repo
3. Select Builder image (node.js for example), enter a name. Now all required fields are filled.

CASE B
1. Switch to developer perspective, add page > import from Git
2. Enter URL https://mysupersecretgit.example.com/org/repo
3. Select Devfile, enter "Devfile", a name. Now all required fields are filled.

**Browser conformance**: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

invincibleJai commented:
/cc @abhinandan13jan

abhinandan13jan commented on frontend/packages/dev-console/src/components/import/devfile/DevfileStrategySection.tsx:
```diff
@@ -77,12 +77,15 @@ const DevfileStrategySection: React.FC = () => {
       return t('devconsole~Validated');
     }
     if (validated === ValidatedOptions.error) {
+      if (type === GitProvider.UNSURE) {
```
How are we certain that UNSURE maps to Devfile always?

abhinandan13jan commented on frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx:
```diff
@@ -0,0 +1,143 @@
+  </Provider>
+);
+
+afterEach(() => cleanup());
```
put this inside describe

lokanandaprabhu replied:
Made the changes

abhinandan13jan commented on frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx:
```diff
@@ -0,0 +1,143 @@
+
+afterEach(() => cleanup());
+
+const devStrategyInput = {
```
nit: You can make this reusable by putting this in external file like frontend/packages/dev-console/integration-tests/testData/git-import-flow.data.ts

lokanandaprabhu replied:
Made the changes

abhinandan13jan commented:
tested locally, works as expected... but I see that revalidation isn't working properly. Not sure whether it is related. cc: @jerolimov 

For devfile flow
![revalidation](https://user-images.githubusercontent.com/24852534/171656423-4afc3165-4f2b-4716-9fca-d18091c514ef.gif)


For github flow
![revalidation2](https://user-images.githubusercontent.com/24852534/171656480-9e9606bf-d5a6-4e19-8066-095604fd67a5.gif)

lokanandaprabhu commented:
> tested again!! the issue around revalidation still exists

I also tested this. My observation is, it is like this in 4.11 (Master) and 4.10 versions also. If URL gets validated successfully (consider for github) it is revalidating for onChange event.

I will wait for @jerolimov response on this.

cc: @jerolimov

jerolimov commented:
/cc
/assign

jerolimov commented on frontend/packages/dev-console/src/components/import/devfile/DevfileStrategySection.tsx:
```diff
@@ -106,9 +109,11 @@ const DevfileStrategySection: React.FC = () => {
 
   React.useEffect(() => {
-    importType === 'devfile' && devfile.devfilePath && handleDevfileChange();
+    (importType === 'devfile' || selectedStrategy.type === ImportStrategy.DEVFILE) &&
```
We have an enum for the importType:

```suggestion
    (importType === ImportTypes.devfile || selectedStrategy.type === ImportStrategy.DEVFILE) &&
```

jerolimov commented on frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx:
```diff
@@ -0,0 +1,89 @@
+  </Provider>
+);
+
+describe('DevFileImportStrategy', () => {
```
Doesn't this test file primary test `DevfileStrategySection` ? In this case the file should also renamed and moved into `../import/devfile/__tests__` folder and should be called `DevfileStrategySection.spec.tsx` or?

```suggestion
describe('DevfileStrategySection', () => {
```

jerolimov commented on frontend/packages/dev-console/integration-tests/testData/git-import-flow.data.ts:
```diff
@@ -13,3 +16,57 @@ export const testData = {
   label: 'app=devlelopment',
 };
+
+export const devStrategyInput = {
```
Should be called devfile

```suggestion
export const devfileStrategyInput = {
```

jerolimov commented on frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx:
```diff
@@ -0,0 +1,89 @@
+import { Provider } from 'react-redux';
+import { GitProvider } from '@console/git-service/src';
+import store from '@console/internal/redux';
+import { devStrategyInput } from '../../../../../integration-tests/testData/git-import-flow.data';
```
When just used in this test case, you can also inline the mock data or create a DevfileStrategySection.data.tsx parallel to this file. These cross imports from a unit test to our e2e make it difficult to maintain this later.

lokanandaprabhu commented:
/retest

lokanandaprabhu commented:
/test e2e-gcp-console

invincibleJai commented on frontend/packages/dev-console/locales/en/devconsole.json:
```diff
@@ -374,6 +374,7 @@
   "Sample repository:": "Sample repository:",
   "The Devfile in your Git repository is invalid.": "The Devfile in your Git repository is invalid.",
   "Validated": "Validated",
+  "Could not get Devfile for an unknown Git type": "Could not get Devfile for an unknown Git type",
```
do we need to check on the wording with UXD/PM?

cc @jerolimov

jerolimov replied:
Sure! Please check with @serenamarie125 or @beaumorley

jerolimov commented:
Hey @lokanandaprabhu , do get this please address the following issues:

1. Rebase this PR
2. Address Jais two review comments (add a type for devfileStrategyInput and use data-test instead of data-test-id) https://github.com/openshift/console/pull/11493#pullrequestreview-1028203535
3. Check with Serena the wording of "Could not get Devfile for an unknown Git type"

When the user selects a strategy and changes the URL (the comment from Abhi) the Import strategy always jumps to Builder Image is IMO not related to this PR. This happens also for GitHub URLs. We could improve this case, but we can address this edge case another time.

jerolimov reviewed (changes requested):
@lokanandaprabhu The recommendation from @serenamarie125 was that we should "disable" the "Devfile" button if the git provider (git type) is other (and we don't support the Devfile).

It's correct that we don't need the user input for a Devfile if we know that this variant is not supported.

We can disable the button and show a tooltip that explains why the button is disabled. Maybe you can share first a draft how this would look like before you update all the tests and cleanup your code.

I'll add hold on this PR so that it's clear that this needs more work before someone reviews this again.

/hold

lokanandaprabhu commented:
Hi @jerolimov ,

I attached the screenshot, if the git type is Other, then Devfile has to be disabled with a tooltip. Just for taking screenshot I added some text there, I can change that. This design is fine or any feedback on this? If this is fine, I will start to work on this.

<img width="795" alt="Screenshot 2022-08-08 at 3 55 38 PM" src="https://user-images.githubusercontent.com/102503482/183407827-eaccddd8-0f85-4987-8259-e3705568b683.png">

lokanandaprabhu commented:
Hi @jerolimov ,

Can you take a look at the comment - https://github.com/openshift/console/pull/11493#issuecomment-1208005946 and provide your feedback. 

Thanks.

lokanandaprabhu commented:
/retest

lokanandaprabhu commented:
/retest

lokanandaprabhu commented:
Hi @jerolimov ,

Requesting your review on this.

jerolimov commented on frontend/packages/dev-console/integration-tests/support/step-definitions/addFlow/create-from-git.ts:
```diff
@@ -256,3 +256,19 @@ Then('user is able to see {string} value is selected in TLS termination', (value
   cy.get(gitPO.advancedOptions.routing.insecureTraffic).should('have.text', value);
 });
+
+When('devfile import strategy is disabled', () => {
```
Nit: This is what you expect:

```suggestion
Then('devfile import strategy is disabled', () => {
```

jerolimov commented:
@lokanandaprabhu Sorry for the delay :scream: I've tested this PR now and it looks good to me. I added one comment for the e2e test above. Can you rebase this PR and then I expect that we can merge it.

jerolimov reviewed (changes requested):
@lokanandaprabhu Sorry for the delay :scream: I've tested this PR now and it looks good to me. I added one comment for the e2e test above. Can you rebase this PR and then I expect that we can merge it.

lokanandaprabhu commented:
> @lokanandaprabhu Sorry for the delay 😱 I've tested this PR now and it looks good to me. I added one comment for the e2e test above. Can you rebase this PR and then I expect that we can merge it.

Hi @jerolimov ,

Updated the PR and rebased it. Thanks.

lokanandaprabhu commented:
/retest

jerolimov commented:
/retest

jerolimov commented:
@lokanandaprabhu the frontend job fails because this linter issue:

```
/go/src/github.com/openshift/console/frontend/packages/dev-console/src/components/import/ImportStrategySelector.tsx
  8:1  error  `@console/git-service/src` import should occur before import of `@console/internal/actions/ui`  import/order
```

jerolimov commented on frontend/packages/dev-console/integration-tests/tsconfig.json:
```diff
@@ -5,7 +5,8 @@
     "esModuleInterop": true,
     "types": ["cypress"],
-    "lib": ["es6", "dom", "es2017"]
+    "lib": ["es6", "dom", "es2017"],
+    "jsx": "react"
```
Is this change really required for this PR?

lokanandaprabhu replied:
Before I have added normal tests not the e2e tests. That time I have added this. Now I removed.

jerolimov commented on frontend/packages/dev-console/src/components/import/git/GitSection.tsx:
```diff
@@ -315,6 +315,7 @@ const GitSection: React.FC<GitSectionProps> = ({
           recommendedStrategy: null,
           showEditImportStrategy: true,
         });
+        setFieldValue('build.strategy', BuildStrategyType.Source);
```
In which case is this needed? The form jumps already on an URL change to the Builder Image (Source) tab, or? But I'm okay with it as I don't see an issue here.

Generally, the form has issues to keep the state that the user has manually touched.

lokanandaprabhu replied:
The above change is for below case

1. Switch to developer perspective, add page > import from Git
2. Enter URL https://mysupersecretgit.example.com/org/repo
3. Select Builder image (node.js for example), enter a name. Now all required fields are filled.

Actual results:
The user can not import the Deployment

Before :


https://user-images.githubusercontent.com/102503482/215422237-2c46437e-62bc-4827-b709-47d57691cb1d.mov

After:

https://user-images.githubusercontent.com/102503482/215422281-1e53b496-fc50-4d31-bf20-e29e751d4b09.mov

jerolimov replied:
Works fine, thank :+1:

jerolimov reviewed (changes requested):
Please check `yarn lint`. Also added two small questions:

jerolimov reviewed (approved):
Works fine :+1: 

/lgtm
/approve

jerolimov commented:
/unhold

jerolimov commented:
/retest

lokanandaprabhu commented:
/retest

jerolimov commented:
/retest

lokanandaprabhu commented:
/retest

lokanandaprabhu commented:
/retest

lokanandaprabhu commented:
/retest

lokanandaprabhu commented:
/retest

lokanandaprabhu commented:
/retest

jerolimov commented:
/unhold
/retest

lokanandaprabhu commented:
/retest

lokanandaprabhu commented:
/unhold
/retest

lokanandaprabhu commented:
/unhold
/retest

jerolimov commented:
/unhold
/retest

lokanandaprabhu commented:
/cherry-pick release-4.12

lokanandaprabhu commented:
manual cherry-pick PR #12775
